minio/api.py

```diff
@@ -1117,34 +1127,49 @@ def stat_object(self, bucket_name, object_name, sse=None):
         :return: Object metadata if object exists
         """
 
+        is_valid_bucket_name(bucket_name, False)
+        is_non_empty_string(object_name)
+
         headers = {}
         if sse:
             is_valid_sse_c_object(sse)
             headers = sse.headers()
 
-        is_valid_bucket_name(bucket_name, False)
-        is_non_empty_string(object_name)
-
-        response = self._url_open('HEAD', bucket_name=bucket_name,
-                                  object_name=object_name, headers=headers)
+        if version_id:
+            if extra_query_params:
+                extra_query_params["versionId"] = version_id
+            else:
+                extra_query_params = {"versionId": version_id}
 
-        etag = response.headers.get('etag', '').replace('"', '')
-        size = int(response.headers.get('content-length', '0'))
-        content_type = response.headers.get('content-type', '')
-        last_modified = response.headers.get('last-modified')
+        response = self._url_open(
+            "HEAD",
+            bucket_name=bucket_name,
+            object_name=object_name,
+            headers=headers,
+            query=extra_query_params,
+        )
 
-        # Capture only custom metadata.
-        custom_metadata = dict()
-        for k in response.headers:
-            if is_supported_header(k) or is_amz_header(k):
-                custom_metadata[k] = response.headers.get(k)
+        custom_metadata = {
+            key: value for key, value in response.headers.items()
+            if is_supported_header(key) or is_amz_header(key)
+        }
 
+        last_modified = response.headers.get("last-modified")
         if last_modified:
             last_modified = dateutil.parser.parse(last_modified).timetuple()
-        return Object(bucket_name, object_name, last_modified, etag, size,
-                      content_type=content_type, metadata=custom_metadata)
 
-    def remove_object(self, bucket_name, object_name):
+        return Object(
+            bucket_name,
+            object_name,
+            last_modified=last_modified,
+            etag=response.headers.get("etag", "").replace('"', ""),
+            size=int(response.headers.get("content-length", "0")),
```

Comment:
I see these headers being canonicalized in s3 docs. Like  `Content-Length` and `Content-Type` etc. Are we cleaning it any place?

looks like all the x-amz-* will be non-canonicalized as per the docs - https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html

Reply:
I think its using `FoldCaseDict` internally. Nothing changed from existing logic.